include/zephyr/drivers/retention.h

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ * @brief Public API for data retention drivers
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_RETENTION_
+#define ZEPHYR_INCLUDE_DRIVERS_RETENTION_
+
+#include <stdint.h>
+#include <stddef.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Data Retention Driver Interface
+ * @defgroup retention_interface Data Retention Driver Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+/**
+ * @typedef	retention_details_api
+ * @brief	
+ * See auxdisplay_custom_command() for argument description
+ */
+//typedef int32_t (*retention_details_api)(const struct device *dev, ); //size of area, start offset, etc.
+
+/**
+ * @typedef	retention_is_valid_api
+ * @brief	Callback API to check if retained data is valid (correct header and checksum,
+ *		if enabled).
+ * See auxdisplay_custom_command() for argument description
+ */
+typedef int32_t (*retention_is_valid_api)(const struct device *dev, bool *valid);
+
+/**
+ * @typedef	retention_read_api
+ * @brief	Callback API to read retained data.
+ * See auxdisplay_custom_command() for argument description
+ */
+typedef int32_t (*retention_read_api)(const struct device *dev, size_t offset, uint8_t *buffer,
+				      size_t *size);
+
+/**
+ * @typedef	retention_set_api
+ * @brief	Callback API to set retained data (will be marked as valid).
+ * See auxdisplay_custom_command() for argument description
+ */
+typedef int32_t (*retention_set_api)(const struct device *dev, size_t offset, uint8_t *buffer,
+				     size_t size);
+
+/**
+ * @typedef	retention_clear_api
+ * @brief	Callback API to clear retained data (will no longer be detected as valid).
+ * See auxdisplay_custom_command() for argument description
+ */
+typedef int32_t (*retention_clear_api)(const struct device *dev);
+
```

Comment:
It seems the implementation of this API is often just memset() to a known value. Unless there is hardware which has some bit or built in operation to set all memory to some preset value, or otherwise mark data as cleared, I would instead replace this API with a memset wrapper instead:
```
typedef int32_t (*retention_memset_api)(const struct device *dev, size_t offset, uint8_t value, size_t size);
```

Reply:
Added clear() function but unimplemented in current drivers as hardware does not support it